### PR TITLE
Add support for hidden branches

### LINKED ISSUE #8355: Support hidden branches feature
In many scenarios, internal lakeFS flows requires creation of ephemeral branches. These branches should not be visible to the users to avoid confusion/friction or accidentally disrupting lakeFS flows (by user modification or deletion).
The suggestion is to allow creation of hidden branches which will not show up when listing repository branches and minimize the exposure of these branches to the end user.

Introduce a new flag for branch creation:
```swagger
    BranchCreation:
      type: object
      required:
        - name
        - source
      properties:
        name:
          type: string
        source:
          type: string
        force:
          type: boolean
          default: false
        hidden:
          type: boolean
          description: when true, branch will not show in list branches by default
          default: false
```

Add `hidden` property to branch which will be defaulted to false if not exists (backwards compatible):
```proto
message BranchData {
  string id = 1;
  string commit_id = 2;
  string staging_token = 3;
  repeated string sealed_tokens = 4;
  bool hidden = 5;
}
```

These branches will not show up when listing repository branches by default. Introduce a new query param to ListBranches command:
```swagger
        - in: query
          name: show_hidden
          schema:
            type: boolean
            default: false
            description: list all branches including hidden branches
```

* Hidden branches will still be available for GetBranch operation regardless of hidden state.
* ATM we will not allow modifying the hidden state of branches

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #8355

### Background

This came as a discussion regarding the use of ephemeral branches. For example for the use of transactions (Python SDK)
      
### New Feature

Add support for creating hidden branches

Hidden branches by default not shown on branch listing
Added a new flag for branch listing to show all branches (disabled by default)
Hidden branches can still be fetched using the GetBranch API

### Testing Details

Add unit testing for graveler and controller

### Breaking Change?

No

### CODE REVIEW

File: pkg/api/controller_test.go
Comment (arielshaqed) [inline pkg/api/controller_test.go:1512]: I think we want a different set of changes to the tests:
- ListBranches should be tested both with and without ShowHidden.  Right now it is not clear where we test ShowHidden=false.
- All other branch tests can just pick one value of Hidden and stick with it.  There is no special behaviour for diff on a hidden branch, or indeed for any other on-branch operation.

File: pkg/api/controller_test.go
Comment (arielshaqed) [inline pkg/api/controller_test.go:1520]: I prefer to compare outputs directly, it makes debugging easier when there are any failures.  So in this case I would extract `.Id`s to a slice, and then use a contents comparison such as deep.Equal (or perhaps testify has a usable slice equality test).

File: api/swagger.yml
Comment (arielshaqed) [inline api/swagger.yml:699]:
```suggestion
        hidden:
          type: boolean
          description: *EXPERIMENTAL* When set, branch will not show up when listing branches by default.          
          default: false
```
(Also below)

**Review by arielshaqed — COMMENTED** [review]
Wow, thanks!  This make isolation on lakeFS so much better!

Small things:

- I would prefer to mark this "experimental".  While I am quite sure that we will want to keep it, I can imagine the API changing.
- Check annotation in catalog_test.go suggests we need to run gomock or something.
- Not sure about controller tests.

Larger issue: I would like us to re-think how we handle "special-purpose" metadata.

I imagine we will end up with user metadata for branches.  Indeed I can think of several purposes for this already.  I think I would prefer to have _general_ metadata for branches -- key-value, essentially.  If we had then, we could declare that we continue to own the `::lakefs::*` namespace (same as we do on objects) -- and then a branch is hidden if it has `::lakefs::hidden` set.

I **do like** the new interface for ListBranches.  If (as an alternative) we implemented ListBranches as "list branches with this general filter language for branch metadata", then hidden branches would not really be hidden unless you asked for that.  So I do _not_ want to do that!

File: pkg/api/controller_test.go
Comment (N-o-Z) [reply pkg/api/controller_test.go:1512]: This is what the test does.
line 1496 - test with hidden=false
line 1509 - test with hidden=true
We have also other unit tests which cover the hidden flag in more variations

File: pkg/api/controller_test.go
Comment (N-o-Z) [reply pkg/api/controller_test.go:1520]: I'm sorry - but this feels like a complete overkill.
We're comparing strings, why would we need deep equals? Also I don't understand how putting the expected and received in a slice and comparing slices makes it easier to debug?
These tests run in ms, so putting a breakpoint where you need and looking and the variables takes less than a second.

**Comment by N-o-Z** [comment]
> Wow, thanks! This make isolation on lakeFS so much better!
> 
> Small things:
> 
> * I would prefer to mark this "experimental".  While I am quite sure that we will want to keep it, I can imagine the API changing.
> * Check annotation in catalog_test.go suggests we need to run gomock or something.
> * Not sure about controller tests.
> 
> Larger issue: I would like us to re-think how we handle "special-purpose" metadata.
> 
> I imagine we will end up with user metadata for branches. Indeed I can think of several purposes for this already. I think I would prefer to have _general_ metadata for branches -- key-value, essentially. If we had then, we could declare that we continue to own the `::lakefs::*` namespace (same as we do on objects) -- and then a branch is hidden if it has `::lakefs::hidden` set.
> 
> I **do like** the new interface for ListBranches. If (as an alternative) we implemented ListBranches as "list branches with this general filter language for branch metadata", then hidden branches would not really be hidden unless you asked for that. So I do _not_ want to do that!

Thanks for the review.
Not so sure about the metadata approach. User metadata is one thing, but defining branch properties as internal metadata is something I think we should be careful with - it's much more prone to errors IMO.
We might want to also refrain from coupling features which don't necessarily have close relation and especially if they don't exist yet (nor on the roadmap).

File: pkg/api/controller_test.go
Comment (arielshaqed) [reply pkg/api/controller_test.go:1512]: Sorry, missed that this is an added test.  Thanks fo explaining!

File: pkg/api/controller_test.go
Comment (arielshaqed) [reply pkg/api/controller_test.go:1520]: Debuggers are the antithesis of testing.
The advantage of comparing slices of strings correctly is that you see entire context.  

- For instance, suppose you receive ["main3"].    You get the useless error message "expected 2 branches to return", and not even the branches you did get.
- For instance, suppose you receive ["main3", "main2"] (maybe some KV weirdness).  You get the useless error message "got main3 expected main2" which still doesn't say what happened.
- For instance, suppose you receive ["main1", "main2"] (some filtering is broken).  You get the useless error message "got main1 expected main2" which still doesn't say what happened.

Tests should fail with a useful error message, if only to help understand how much work is required to fix.

**Review by arielshaqed — APPROVED** [review]
Thanks!  Asking for better tests but not requiring it for approval.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Waiting for product feedback

File: pkg/api/controller_test.go
Comment (N-o-Z) [reply pkg/api/controller_test.go]: modified

**Pull request merged**
